### PR TITLE
Release/v3.0.0

### PR DESCRIPTION
**Pull request by CiaranOMara**
This PR refactors `GenomicFeatures` to make use of abstract types.

This PR contains the following breaking changes.
- The `Interval` type is renamed `GenomicInterval` and now subtypes `AbstractGenomicInterval`.
- The  `IntervalCollection` is now `GenomicIntervalCollection`, and the element type of the collection is now the `AbstractGenomicInterval` instead of the interval's metadata type.
- Intervals are now grouped by the more generic `groupname` instead of `seqname`.

### CODE REVIEW

**Review by kescobo — COMMENTED** [review]
I'm not going to pretend I understand the scope of these changes. I can do a more detailed review if you think that would be helpful, but I trust you to make good decisions here. 

My only high-level comment is that all doctests should probably pass (or be removed) at this stage before merging. No reason to have our first post-1.0 docs be instantly outdated :wink:

**Comment by CiaranOMara** [comment]
@kescobo, this PR is now ready for review again. I think I addressed your comment and tidied up a few loose ends.

**Review by kescobo — CHANGES_REQUESTED** [review]
I left a bunch of comments, but the vast majority amount to style preferences, and I do not feel strongly enough to consider them blocking. There is one spot where it looks like a `T` didn't get correctly changed to `I` (or vice versa), but that's the only part of this review that I would insist on.

The big style ones that show up in a few places look as if you were copying the previous style, but I think should be changed are

1. A bunch of block that have the struture `if ... return X; end; if ... return Y; end; return Z` that I think should either be `if/elseif/else`, or else use short-circuiting `&& return X`
2. A bunch of function definitions that are are just returning something that I think could be written with `foo() = X` rather than `function foo(); return X; end`

In both cases, I commented on one or two examples, but there were many more that I didn't flag. If you do want to make these changes but don't want to put in the extra tedious effort, I can do them as a separate PR - neither of them should change the functionality, so they can safely be done after v3.0 is tagged.

File: src/GenomicFeatures.jl
Comment (kescobo) [reply src/GenomicFeatures.jl:92]: @CiaranOMara This is the one thing that I think *needs* a change. Everything else is style, I think

Comment (CiaranOMara) [reply src/GenomicFeatures.jl:92]: That definitely needs to change. That it got this far undetected definitely reinforces a need for test coverage. I've added tests to the commit in which it was introduced.

File: src/GenomicFeatures.jl
Comment (CiaranOMara) [reply src/GenomicFeatures.jl:111]: My preference is to use the verbose function block because I've found IDE hyperclicks unreliable or slower than a global search.

File: src/coverage.jl
Comment (CiaranOMara) [reply src/coverage.jl:121]: Not sure. This use of `UInt32` was a historic choice made by either @bicycle1885 or @SabrinaJaye. We can loosen the type requirement after this is merged and check the benchmarks.

File: src/interval.jl
Comment (CiaranOMara) [reply src/interval.jl]: You're correct. Parentheses removed.

File: src/GenomicFeatures.jl
Comment (kescobo) [reply src/GenomicFeatures.jl:107]: OK, 100% your call

**Review by kescobo — DISMISSED** [review]

**Comment by CiaranOMara** [comment]
Thanks @kescobo.

**Pull request merged**
